### ISSUE TITLE
NSA consumes longer time to train than FULL?

### ISSUE THREAD

**Issue by GeneZC7** [mdy666/qwen-nsa#1]
From the loss figure, it seems NSA takes longer time to consume the same number of samples than FULL, why?

https://github.com/mdy666/qwen-nsa/blob/master/log/imgs/loss-7B.png

BTW, according to the original paper, there should be fixed initial and local blocks selected, which is not considered in this implementation.

https://github.com/mdy666/qwen-nsa/blob/master/nsa_attention/select_attn_v3.py#L143

**Comment by mdy666** [comment]
> From the loss figure, it seems NSA takes longer time to consume the same number of samples than FULL, why?
> you can see the benchmark, 4k seq_len is slower than fa2, If pretrain in 8k, the speed close the fa2. The seq_len is more long the speed up is more obviously.  But in fact, the seq_len mostly be 4k or 8k in the pretraining phase.

> https://github.com/mdy666/qwen-nsa/blob/master/log/imgs/loss-7B.png
> 
> BTW, according to the original paper, there should be fixed initial and local blocks selected, which is not considered in this implementation.
> I think the "fixed initial" means the query must select the block which it local, for example if the q_idx = 80, it must select the block which the scope between 64 and 128, you can see the figure in the paper. And "local blocks" means other block
> https://github.com/mdy666/qwen-nsa/blob/master/nsa_attention/select_attn_v3.py#L143

**Comment by GeneZC7** [comment]
I see, thanks.

And I want to highlight that the initial blocks may mean something like block-wise attention sinks if I understand correctly. The local blocks indeed mean the thing you consider.

Closing since no further questions~
